File: src/main/java/org/moinex/service/WalletService.java
/*
 * Filename: WalletService.java
 * Created on: August 31, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.service;

import jakarta.persistence.EntityExistsException;
import jakarta.persistence.EntityNotFoundException;
import java.math.BigDecimal;
import java.util.Comparator;
import java.util.List;
import lombok.NoArgsConstructor;
import lombok.NonNull;
import org.moinex.error.MoinexException;
import org.moinex.model.wallettransaction.Wallet;
import org.moinex.model.wallettransaction.WalletType;
import org.moinex.repository.goal.GoalRepository;
import org.moinex.repository.wallettransaction.TransferRepository;
import org.moinex.repository.wallettransaction.WalletRepository;
import org.moinex.repository.wallettransaction.WalletTransactionRepository;
import org.moinex.repository.wallettransaction.WalletTypeRepository;
import org.moinex.util.UIUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * This class is responsible for the business logic of the Wallet entity
 */
@Service
@NoArgsConstructor
public class WalletService {
    private static final Logger logger = LoggerFactory.getLogger(WalletService.class);
    private WalletRepository walletRepository;
    private TransferRepository transfersRepository;
    private WalletTransactionRepository walletTransactionRepository;
    private WalletTypeRepository walletTypeRepository;
    private GoalRepository goalRepository;

    @Autowired
    public WalletService(
            WalletRepository walletRepository,
            TransferRepository transfersRepository,
            WalletTransactionRepository walletTransactionRepository,
            WalletTypeRepository walletTypeRepository,
            GoalRepository goalRepository) {
        this.walletRepository = walletRepository;
        this.transfersRepository = transfersRepository;
        this.walletTransactionRepository = walletTransactionRepository;
        this.walletTypeRepository = walletTypeRepository;
        this.goalRepository = goalRepository;
    }

    /**
     * Creates a new wallet
     *
     * @param name       The name of the wallet
     * @param balance    The initial balance of the wallet
     * @param walletType The type of the wallet
     * @return The id of the created wallet
     * @throws IllegalArgumentException If the wallet name is empty
     * @throws EntityExistsException    If the wallet name is already in use
     */
    @Transactional
    public Integer addWallet(String name, BigDecimal balance, WalletType walletType) {
        // Remove leading and trailing whitespaces
        name = name.strip();

        if (name.isBlank()) {
            throw new IllegalArgumentException("Wallet name cannot be empty");
        }

        if (walletRepository.existsByName(name)) {
            throw new EntityExistsException("Wallet with name " + name + " already exists");
        }

        logger.info("Wallet {} created with balance {}", name, balance);

        Wallet wt = Wallet.builder().name(name).balance(balance).type(walletType).build();

        walletRepository.save(wt);

        return wt.getId();
    }

    /**
     * Delete a wallet
     *
     * @param id The id of the wallet to be deleted
     * @throws EntityNotFoundException If the wallet does not exist
     * @throws IllegalStateException   If the wallet has transactions
     */
    @Transactional
    public void deleteWallet(Integer id) {
        Wallet wallet =
                walletRepository
                        .findById(id)
                        .orElseThrow(
                                () ->
                                        new EntityNotFoundException(
                                                String.format(
                                                        "Wallet with id %d not found and cannot be"
                                                                + " deleted",
                                                        id)));

        if (walletTransactionRepository.getTransactionCountByWallet(id) > 0
                || transfersRepository.getTransferCountByWallet(id) > 0) {
            throw new IllegalStateException(
                    String.format(
                            "Wallet with id %d has transactions and cannot be deleted."
                                    + "Remove the transactions first or archive the wallet",
                            id));
        }

        if (wallet.isMaster()) {
            removeAllVirtualWalletsFromMasterWallet(wallet.getId());
        }

        walletRepository.delete(wallet);

        logger.info("Wallet with id {} was permanently deleted", id);
    }

    public void removeAllVirtualWalletsFromMasterWallet(Integer masterWalletId) {
        List<Wallet> virtualWallets =
                walletRepository.findVirtualWalletsByMasterWallet(masterWalletId);

        if (!virtualWallets.isEmpty()) {
            logger.info(
                    "Removing {} virtual wallets from master wallet with id {}",
                    virtualWallets.size(),
                    masterWalletId);
            virtualWallets.forEach(
                    virtualWallet -> {
                        virtualWallet.setMasterWallet(null);
                        walletRepository.save(virtualWallet);
                        logger.info(
                                "Virtual wallet with id {} unlinked from master wallet",
                                virtualWallet.getId());
                    });
        }
    }

    /**
     * Archive a wallet
     *
     * @param id The id of the wallet to be archived
     * @throws EntityNotFoundException If the wallet does not exist
     * @note This method is used to archive a wallet, which means that the wallet
     * will not be deleted from the database, but it will not be used in the
     * application anymore
     */
    @Transactional
    public void archiveWallet(Integer id) {
        Wallet wallet =
                walletRepository
                        .findById(id)
                        .orElseThrow(
                                () ->
                                        new EntityNotFoundException(
                                                String.format(
                                                        "Wallet with id %d not found and cannot be"
                                                                + " archived",
                                                        id)));
        wallet.setArchived(true);

        if (wallet.isMaster()) {
            updateWalletBalance(wallet.getId(), getUnallocatedBalance(wallet));
            removeAllVirtualWalletsFromMasterWallet(wallet.getId());
        } else if (wallet.isVirtual()) {
            wallet.setMasterWallet(null);
        }

        walletRepository.save(wallet);

        logger.info("Wallet with id {} was archived", id);
    }

    /**
     * Unarchive a wallet
     *
     * @param id The id of the wallet to be unarchived
     * @throws EntityNotFoundException If the wallet does not exist
     * @note This method is used to unarchive a wallet, which means that the wallet
     * will be used in the application again
     */
    @Transactional
    public void unarchiveWallet(Integer id) {
        Wallet wallet =
                walletRepository
                        .findById(id)
                        .orElseThrow(
                                () ->
                                        new EntityNotFoundException(
                                                String.format(
                                                        "Wallet with id %d not found and cannot be"
                                                                + " unarchived",
                                                        id)));

        wallet.setArchived(false);
        walletRepository.save(wallet);

        logger.info("Wallet with id {} was unarchived", id);
    }

    /**
     * Rename a wallet
     *
     * @param id      The id of the wallet to be renamed
     * @param newName The new name of the wallet
     * @throws IllegalArgumentException If the wallet name is empty
     * @throws EntityNotFoundException  If the wallet does not exist
     * @throws EntityExistsException    If the wallet name is already in use
     */
    @Transactional
    public void renameWallet(Integer id, String newName) {
        // Remove leading and trailing whitespaces
        newName = newName.strip();

        if (newName.isBlank()) {
            throw new IllegalArgumentException("Wallet name cannot be empty");
        }

        Wallet wallet =
                walletRepository
                        .findById(id)
                        .orElseThrow(
                                () ->
                                        new EntityNotFoundException(
                                                String.format("Wallet with id %d not found", id)));

        if (walletRepository.existsByName(newName)) {
            throw new EntityExistsException("Wallet with name " + newName + " already exists");
        }

        wallet.setName(newName);
        walletRepository.save(wallet);

        logger.info("Wallet with id {} renamed to {}", id, newName);
    }

    /**
     * Change the wallet type
     *
     * @param id      The id of the wallet to change the type
     * @param newType The new type of the wallet
     * @throws EntityNotFoundException                      If the wallet does not exist
     * @throws EntityNotFoundException                      If the wallet type does not exist
     * @throws MoinexException.AttributeAlreadySetException If the wallet already has the new type
     */
    @Transactional
    public void changeWalletType(Integer id, WalletType newType) {
        Wallet wallet =
                walletRepository
                        .findById(id)
                        .orElseThrow(
                                () ->
                                        new EntityNotFoundException(
                                                String.format("Wallet with id %d not found", id)));

        if (newType == null || !walletTypeRepository.existsById(newType.getId())) {
            throw new EntityNotFoundException("Wallet type not found");
        }

        if (wallet.getType().getId().equals(newType.getId())) {
            throw new MoinexException.AttributeAlreadySetException(
                    "Wallet with name "
                            + wallet.getName()
                            + " already has type "
                            + newType.getName());
        }

        wallet.setType(newType);
        walletRepository.save(wallet);

        logger.info("Wallet with id {} type changed to {}", id, newType.getName());
    }

    /**
     * Update the balance of a wallet
     *
     * @param id         The id of the wallet
     * @param newBalance The new balance of the wallet
     * @throws EntityNotFoundException If the wallet does not exist
     */
    @Transactional
    public void updateWalletBalance(Integer id, BigDecimal newBalance) {
        Wallet wallet =
                walletRepository
                        .findById(id)
                        .orElseThrow(
                                () ->
                                        new EntityNotFoundException(
                                                String.format("Wallet with id %d not found", id)));

        BigDecimal diff = newBalance.subtract(wallet.getBalance());

        if (diff.compareTo(BigDecimal.ZERO) > 0) {
            updateWalletBalance(id, diff, true);
        } else if (diff.compareTo(BigDecimal.ZERO) < 0) {
            updateWalletBalance(id, diff.negate(), false);
        } else {
            logger.info("Wallet with id {} balance remains unchanged", id);
        }
    }

    @Transactional
    public void incrementWalletBalance(Integer id, BigDecimal amount) {
        updateWalletBalance(id, amount, true);
    }

    @Transactional
    public void decrementWalletBalance(Integer id, BigDecimal amount) {
        updateWalletBalance(id, amount, false);
    }

    private void updateWalletBalance(
            @NonNull Integer id, @NonNull BigDecimal amount, boolean isIncrement) {
        if (amount.compareTo(BigDecimal.ZERO) <= 0) {
            throw new IllegalArgumentException("Amount must be positive");
        }

        Wallet wallet =
                walletRepository
                        .findById(id)
                        .orElseThrow(
                                () ->
                                        new EntityNotFoundException(
                                                String.format("Wallet with id %d not found", id)));

        BigDecimal adjustedAmount = isIncrement ? amount : amount.negate();

        if (wallet.isVirtual()) {
            if (wallet.getBalance().add(adjustedAmount).compareTo(BigDecimal.ZERO) < 0) {
                throw new IllegalStateException("Virtual wallets cannot have negative balances");
            }

            Wallet master = wallet.getMasterWallet();
            master.setBalance(master.getBalance().add(adjustedAmount));
            walletRepository.save(master);

            logger.info(
                    "Master wallet with id {} balance {} by {}",
                    master.getId(),
                    isIncrement ? "incremented" : "decremented",
                    amount);

        } else if (wallet.isMaster() && !isIncrement) {
            BigDecimal unallocatedBalance = getUnallocatedBalance(wallet);

            if (unallocatedBalance.compareTo(amount) < 0) {
                throw new IllegalStateException(
                        "Master wallet cannot be decremented below the unallocated balance. "
                                + "\nUnallocated balance is "
                                + UIUtils.formatCurrency(unallocatedBalance));
            }
        }

        wallet.setBalance(wallet.getBalance().add(adjustedAmount));
        walletRepository.save(wallet);

        logger.info(
                "Wallet with id {} balance {} by {}",
                wallet.getId(),
                isIncrement ? "incremented" : "decremented",
                amount);
    }

    /**
     * Get wallet by id
     *
     * @param id The id of the wallet
     * @return The wallet with the provided id
     * @throws EntityNotFoundException If the wallet does not exist
     */
    public Wallet getWalletById(Integer id) {
        return walletRepository
                .findById(id)
                .orElseThrow(
                        () -> new EntityNotFoundException("Wallet with id " + id + " not found"));
    }

    /**
     * Get all wallet types
     *
     * @return A list with all wallet types
     */
    public List<WalletType> getAllWalletTypes() {
        return walletTypeRepository.findAllByOrderByNameAsc();
    }

    /**
     * Get all archived wallets
     *
     * @return A list with all archived wallets
     */
    public List<Wallet> getAllArchivedWallets() {
        return walletRepository.findAllByIsArchivedTrue();
    }

    /**
     * Get all wallets that are not archived ordered by name
     *
     * @return A list with all wallets that are not archived
     */
    public List<Wallet> getAllNonArchivedWalletsOrderedByName() {
        return walletRepository.findAllByIsArchivedFalseOrderByNameAsc();
    }

    /**
     * Get all wallets that are not archived ordered descending by the number of
     * transactions
     *
     * @return A list with all wallets that are not archived
     */
    public List<Wallet> getAllNonArchivedWalletsOrderedByTransactionCountDesc() {
        return walletRepository.findAllByIsArchivedFalse().stream()
                .sorted(
                        Comparator.comparingLong(
                                        (Wallet w) ->
                                                walletTransactionRepository
                                                                .getTransactionCountByWallet(
                                                                        w.getId())
                                                        + transfersRepository
                                                                .getTransferCountByWallet(
                                                                        w.getId()))
                                .reversed())
                .toList();
    }

    /**
     * Get the unallocated balance of a master wallet.
     * This is the balance that is not allocated to any goals
     *
     * @param masterWallet The master wallet to check
     * @return The unallocated balance of the master wallet
     */
    public BigDecimal getUnallocatedBalance(@NonNull Wallet masterWallet) {
        BigDecimal allocatedBalance =
                walletRepository.getAllocatedBalanceByMasterWallet(masterWallet.getId());

        return masterWallet.getBalance().subtract(allocatedBalance);
    }

    public Integer getCountOfVirtualWalletsByMasterWalletId(Integer masterWalletId) {
        return walletRepository.getCountOfVirtualWalletsByMasterWalletId(masterWalletId);
    }
}

File: src/main/kotlin/org/moinex/service/WalletService.kt
/*
 * Filename: WalletService.kt (original filename: WalletService.java)
 * Created on: August 31, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 03/04/2026
 */

package org.moinex.service

import org.moinex.common.findByIdOrThrow
import org.moinex.common.isZero
import org.moinex.model.enums.WalletTransactionStatus
import org.moinex.model.enums.WalletTransactionType
import org.moinex.model.wallettransaction.Transfer
import org.moinex.model.wallettransaction.Wallet
import org.moinex.model.wallettransaction.WalletTransaction
import org.moinex.model.wallettransaction.WalletType
import org.moinex.repository.wallettransaction.TransferRepository
import org.moinex.repository.wallettransaction.WalletRepository
import org.moinex.repository.wallettransaction.WalletTransactionRepository
import org.moinex.repository.wallettransaction.WalletTypeRepository
import org.moinex.util.UIUtils
import org.slf4j.LoggerFactory
import org.springframework.data.domain.PageRequest
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.math.BigDecimal
import java.time.LocalDateTime
import java.time.Year
import java.time.YearMonth

@Service
class WalletService(
    private val walletRepository: WalletRepository,
    private val transfersRepository: TransferRepository,
    private val walletTransactionRepository: WalletTransactionRepository,
    private val walletTypeRepository: WalletTypeRepository,
) {
    private val logger = LoggerFactory.getLogger(WalletService::class.java)

    @Transactional
    fun createWallet(wallet: Wallet): Int {
        check(!walletRepository.existsByName(wallet.name)) {
            "Wallet with name '${wallet.name}' already exists"
        }

        val newWallet = walletRepository.save(wallet)

        logger.info("$newWallet created successfully")

        return newWallet.id!!
    }

    @Transactional
    fun deleteWallet(id: Int) {
        val walletFromDatabase = walletRepository.findByIdOrThrow(id)

        check(getWalletTransactionAndTransferCountByWallet(id) == 0) {
            "$walletFromDatabase has transactions and cannot be deleted. " +
                "Remove the transactions first or archive the wallet"
        }

        if (walletFromDatabase.isMaster()) {
            removeAllVirtualWalletsFromMasterWallet(walletFromDatabase.id!!)
        }

        walletRepository.delete(walletFromDatabase)

        logger.info("Wallet with id {} was permanently deleted", id)
    }

    @Transactional
    fun archiveWallet(id: Int) {
        val walletFromDatabase = walletRepository.findByIdOrThrow(id)

        walletFromDatabase.apply {
            isArchived = true
        }

        if (walletFromDatabase.isMaster()) {
            walletFromDatabase.apply { balance = getUnallocatedBalance(walletFromDatabase) }
            updateWalletBalance(walletFromDatabase)
            removeAllVirtualWalletsFromMasterWallet(walletFromDatabase.id!!)
        } else {
            walletFromDatabase.masterWallet = null
        }

        logger.info("$walletFromDatabase archived successfully")
    }

    @Transactional
    fun unarchiveWallet(id: Int) {
        val walletFromDatabase = walletRepository.findByIdOrThrow(id)

        walletFromDatabase.apply {
            isArchived = false
        }

        logger.info("$walletFromDatabase unarchived successfully")
    }

    @Transactional
    fun renameWallet(updatedWallet: Wallet) {
        val walletFromDatabase = walletRepository.findByIdOrThrow(updatedWallet.id!!)

        check(!walletRepository.existsByNameAndIdNot(updatedWallet.name, updatedWallet.id!!)) {
            "Wallet with name '${updatedWallet.name}' already exists"
        }

        walletFromDatabase.apply {
            name = updatedWallet.name
        }

        logger.info("$walletFromDatabase renamed successfully")
    }

    @Transactional
    fun changeWalletType(updatedWallet: Wallet) {
        val walletFromDatabase = walletRepository.findByIdOrThrow(updatedWallet.id!!)

        check(walletTypeRepository.existsById(updatedWallet.type.id!!)) {
            "Wallet type '${updatedWallet.type.name}' does not exist"
        }

        walletFromDatabase.apply {
            type = updatedWallet.type
        }

        logger.info("$walletFromDatabase type changed successfully")
    }

    @Transactional
    fun updateWalletBalance(updatedWallet: Wallet) {
        val walletFromDatabase = walletRepository.findByIdOrThrow(updatedWallet.id!!)

        val diff = updatedWallet.balance.subtract(walletFromDatabase.balance)

        when {
            diff > BigDecimal.ZERO -> incrementWalletBalance(walletFromDatabase, diff)
            diff < BigDecimal.ZERO -> decrementWalletBalance(walletFromDatabase, diff.abs())
            else -> logger.info("$walletFromDatabase balance remains unchanged")
        }
    }

    @Transactional
    fun removeAllVirtualWalletsFromMasterWallet(masterWalletId: Int) {
        val virtualWallets = walletRepository.findVirtualWalletsByMasterWallet(masterWalletId)

        virtualWallets.forEach { virtualWallet ->
            virtualWallet.masterWallet = null
            logger.info(
                "Virtual wallet with id {} unlinked from master wallet with id {}",
                virtualWallet.id,
                masterWalletId,
            )
        }
    }

    @Transactional
    fun createWalletTransaction(transaction: WalletTransaction): Int {
        val walletFromDatabase = walletRepository.findByIdOrThrow(transaction.wallet.id!!)

        val newTransaction = walletTransactionRepository.save(transaction)

        if (transaction.status == WalletTransactionStatus.CONFIRMED) {
            when (transaction.type) {
                WalletTransactionType.INCOME -> incrementWalletBalance(walletFromDatabase, transaction.amount)
                WalletTransactionType.EXPENSE -> decrementWalletBalance(walletFromDatabase, transaction.amount)
            }
        }

        logger.info(
            "${transaction.type} with status ${transaction.status} of ${UIUtils.formatCurrency(
                transaction.amount,
            )} added to wallet with id ${walletFromDatabase.id}",
        )

        return newTransaction.id!!
    }

    @Transactional
    fun updateWalletTransaction(updatedTransaction: WalletTransaction) {
        val transactionFromDatabase = walletTransactionRepository.findByIdOrThrow(updatedTransaction.id!!)

        updateTransactionWallet(transactionFromDatabase, updatedTransaction.wallet)
        updateTransactionType(transactionFromDatabase, updatedTransaction.type)
        updateTransactionAmount(transactionFromDatabase, updatedTransaction.amount)
        updateTransactionStatus(transactionFromDatabase, updatedTransaction.status)

        transactionFromDatabase.apply {
            date = updatedTransaction.date
            description = updatedTransaction.description
            category = updatedTransaction.category
            includeInAnalysis = updatedTransaction.includeInAnalysis
        }

        logger.info("$transactionFromDatabase updated successfully")
    }

    @Transactional
    fun deleteWalletTransaction(transactionId: Int) {
        val transactionFromDatabase = walletTransactionRepository.findByIdOrThrow(transactionId)

        if (transactionFromDatabase.status == WalletTransactionStatus.CONFIRMED) {
            when (transactionFromDatabase.type) {
                WalletTransactionType.INCOME -> decrementWalletBalance(transactionFromDatabase.wallet, transactionFromDatabase.amount)
                WalletTransactionType.EXPENSE -> incrementWalletBalance(transactionFromDatabase.wallet, transactionFromDatabase.amount)
            }
        }

        walletTransactionRepository.delete(transactionFromDatabase)

        logger.info("$transactionFromDatabase deleted from ${transactionFromDatabase.wallet}")
    }

    @Transactional
    fun createTransfer(transfer: Transfer): Int {
        val senderWallet = walletRepository.findByIdOrThrow(transfer.senderWallet.id!!)
        val receiverWallet = walletRepository.findByIdOrThrow(transfer.receiverWallet.id!!)

        check(!(senderWallet.isMaster() && receiverWallet.isVirtual() && receiverWallet.masterWallet == senderWallet)) {
            "You cannot transfer money from a master wallet to its virtual wallet."
        }

        check(senderWallet.balance >= transfer.amount) {
            "Sender wallet does not have enough balance to transfer"
        }

        val newTransfer = transfersRepository.save(transfer)

        decrementWalletBalance(senderWallet, transfer.amount)
        incrementWalletBalance(receiverWallet, transfer.amount)

        logger.info("$newTransfer was created")

        return newTransfer.id!!
    }

    @Transactional
    fun updateTransfer(updatedTransfer: Transfer) {
        val transferFromDatabase = transfersRepository.findByIdOrThrow(updatedTransfer.id!!)

        revertTransferWalletBalances(transferFromDatabase, updatedTransfer)

        transferFromDatabase.apply {
            senderWallet = updatedTransfer.senderWallet
            receiverWallet = updatedTransfer.receiverWallet
            amount = updatedTransfer.amount
            date = updatedTransfer.date
            description = updatedTransfer.description
            category = updatedTransfer.category
        }

        logger.info("$transferFromDatabase updated successfully")
    }

    @Transactional
    fun deleteTransfer(transferId: Int) {
        val transferFromDatabase = transfersRepository.findByIdOrThrow(transferId)

        val senderWallet = transferFromDatabase.senderWallet
        val receiverWallet = transferFromDatabase.receiverWallet
        val amount = transferFromDatabase.amount

        logger.info("Reverting $transferFromDatabase")

        incrementWalletBalance(senderWallet, amount)
        decrementWalletBalance(receiverWallet, amount)

        transfersRepository.delete(transferFromDatabase)

        logger.info("$transferFromDatabase deleted successfully")
    }

    fun getWalletById(id: Int): Wallet = walletRepository.findByIdOrThrow(id)

    fun getWalletTransactionById(id: Int): WalletTransaction = walletTransactionRepository.findByIdOrThrow(id)

    fun getAllWalletTypes(): List<WalletType> = walletTypeRepository.findAllByOrderByNameAsc()

    fun existsWalletTypeByName(name: String): Boolean = walletTypeRepository.existsByName(name)

    fun getAllArchivedWallets(): List<Wallet> = walletRepository.findAllByIsArchivedTrue()

    fun getAllNonArchivedWalletsOrderedByName(): List<Wallet> = walletRepository.findAllByIsArchivedFalseOrderByNameAsc()

    fun getAllWalletsOrderedByName(): List<Wallet> = walletRepository.findAllByOrderByNameAsc()

    fun getAllNonArchivedWalletsOrderedByTransactionCountDesc(): List<Wallet> =
        walletRepository
            .findAllByIsArchivedFalse()
            .sortedByDescending { wallet ->
                walletTransactionRepository.getTransactionCountByWallet(wallet.id!!) +
                    transfersRepository.getTransferCountByWallet(wallet.id!!)
            }

    fun getUnallocatedBalance(masterWallet: Wallet): BigDecimal {
        val allocatedBalance =
            walletRepository.getAllocatedBalanceByMasterWallet(masterWallet.id!!)

        return masterWallet.balance.subtract(allocatedBalance)
    }

    fun getCountOfVirtualWalletsByMasterWalletId(masterWalletId: Int): Int =
        walletRepository.getCountOfVirtualWalletsByMasterWalletId(masterWalletId)

    fun getAllNonArchivedWalletTransactionsByType(type: WalletTransactionType): List<WalletTransaction> =
        walletTransactionRepository.findAllNonArchivedTransactionsByType(type)

    fun getAllWalletTransactionsByMonth(yearMonth: YearMonth): List<WalletTransaction> =
        walletTransactionRepository.findTransactionsByMonth(yearMonth.monthValue, yearMonth.year)

    fun getAllWalletTransactionsByWalletAndMonth(
        walletId: Int,
        yearMonth: YearMonth,
    ): List<WalletTransaction> =
        walletTransactionRepository.findTransactionsByWalletAndMonth(
            walletId,
            yearMonth.monthValue,
            yearMonth.year,
        )

    fun getAllNonArchivedWalletTransactionsByMonth(yearMonth: YearMonth): List<WalletTransaction> =
        walletTransactionRepository.findNonArchivedTransactionsByMonth(yearMonth.monthValue, yearMonth.year)

    fun getAllNonArchivedWalletTransactionsByMonthForAnalysis(yearMonth: YearMonth): List<WalletTransaction> =
        walletTransactionRepository.findNonArchivedTransactionsByMonthForAnalysis(yearMonth.monthValue, yearMonth.year)

    fun getAllNonArchivedWalletTransactionsByYear(year: Year): List<WalletTransaction> =
        walletTransactionRepository.findNonArchivedTransactionsByYear(year.value)

    fun getAllNonArchivedWalletTransactionsByWalletAndMonth(
        walletId: Int,
        yearMonth: YearMonth,
    ): List<WalletTransaction> =
        walletTransactionRepository.findNonArchivedTransactionsByWalletAndMonth(
            walletId,
            yearMonth.monthValue,
            yearMonth.year,
        )

    fun getAllNonArchivedWalletTransactionsBetweenDates(
        startDate: LocalDateTime,
        endDate: LocalDateTime,
    ): List<WalletTransaction> = walletTransactionRepository.findNonArchivedTransactionsBetweenDates(startDate, endDate)

    fun getAllNonArchivedConfirmedWalletTransactionsByMonth(yearMonth: YearMonth): List<WalletTransaction> =
        walletTransactionRepository.findNonArchivedConfirmedTransactionsByMonth(
            yearMonth.monthValue,
            yearMonth.year,
        )

    fun getAllNonArchivedLastWalletTransactions(n: Int): List<WalletTransaction> =
        walletTransactionRepository.findNonArchivedLastTransactions(
            PageRequest.ofSize(n),
        )

    fun getOldestWalletTransactionDate(): LocalDateTime = walletTransactionRepository.findOldestTransactionDate() ?: LocalDateTime.now()

    fun getWalletTransactionAndTransferCountByWallet(walletId: Int): Int =
        walletTransactionRepository.getTransactionCountByWallet(walletId) +
            transfersRepository.getTransferCountByWallet(walletId)

    fun getWalletTransactionAndTransferCountByCategory(categoryId: Int): Int =
        walletTransactionRepository.getTransactionCountByCategory(categoryId) +
            transfersRepository.getTransferCountByCategory(categoryId)

    fun getWalletTransactionSuggestionsByType(type: WalletTransactionType): List<WalletTransaction> =
        walletTransactionRepository.findSuggestions(type)

    fun getAllWalletTransactionsByWalletAfterDate(
        walletId: Int,
        date: LocalDateTime,
    ): List<WalletTransaction> = walletTransactionRepository.findTransactionsByWalletAfterDate(walletId, date)

    fun getFirstWalletTransactionDate(walletId: Int): LocalDateTime? = walletTransactionRepository.findFirstTransactionDate(walletId)

    fun getAllTransfers(): List<Transfer> = transfersRepository.findAll()

    fun getTransfersByWalletAndMonth(
        walletId: Int,
        yearMonth: YearMonth,
    ): List<Transfer> = transfersRepository.findTransfersByWalletAndMonth(walletId, yearMonth.monthValue, yearMonth.year)

    fun getTransferSuggestions(): List<Transfer> = transfersRepository.findSuggestions()

    fun existsByName(name: String): Boolean = walletRepository.existsByName(name)

    fun existsByNameAndIdNot(
        name: String,
        id: Int,
    ): Boolean = walletRepository.existsByNameAndIdNot(name, id)

    private fun incrementWalletBalance(
        walletFromDatabase: Wallet,
        amount: BigDecimal,
    ) {
        updateWalletBalance(walletFromDatabase, amount, true)
    }

    private fun decrementWalletBalance(
        walletFromDatabase: Wallet,
        amount: BigDecimal,
    ) {
        updateWalletBalance(walletFromDatabase, amount, false)
    }

    private fun updateWalletBalance(
        walletToUpdate: Wallet,
        amount: BigDecimal,
        isIncrement: Boolean,
    ) {
        if (amount.isZero()) return

        check(amount > BigDecimal.ZERO) {
            "Amount must be positive but was $amount"
        }

        val adjustedAmount = if (isIncrement) amount else amount.negate()

        when {
            walletToUpdate.isVirtual() -> {
                check(walletToUpdate.balance.add(adjustedAmount) >= BigDecimal.ZERO) {
                    "Virtual wallets cannot have negative balances"
                }

                val master = walletToUpdate.masterWallet!!
                master.balance += adjustedAmount

                logger.info(
                    "{} (master) balance {} by {}",
                    master,
                    if (isIncrement) "incremented" else "decremented",
                    amount,
                )
            }
            walletToUpdate.isMaster() && !isIncrement -> {
                val unallocatedBalance = getUnallocatedBalance(walletToUpdate)

                check(unallocatedBalance >= amount) {
                    "Master wallet cannot be decremented below the unallocated balance. " +
                        "\nUnallocated balance is ${UIUtils.formatCurrency(unallocatedBalance)}"
                }
            }
        }

        walletToUpdate.balance += adjustedAmount

        logger.info(
            "Wallet with id {} balance {} by {}",
            walletToUpdate.id,
            if (isIncrement) "incremented" else "decremented",
            amount,
        )
    }

    private fun updateTransactionWallet(
        oldTransaction: WalletTransaction,
        newWallet: Wallet,
    ) {
        if (oldTransaction.wallet.id == newWallet.id) {
            logger.info("Transaction with id {} has the same wallet as before", oldTransaction.id)
            return
        }

        if (oldTransaction.status == WalletTransactionStatus.CONFIRMED) {
            when (oldTransaction.type) {
                WalletTransactionType.EXPENSE -> {
                    incrementWalletBalance(oldTransaction.wallet, oldTransaction.amount)
                    decrementWalletBalance(newWallet, oldTransaction.amount)
                }
                WalletTransactionType.INCOME -> {
                    decrementWalletBalance(oldTransaction.wallet, oldTransaction.amount)
                    incrementWalletBalance(newWallet, oldTransaction.amount)
                }
            }
        }

        oldTransaction.wallet = newWallet

        logger.info(
            "Transaction with id {} wallet changed to {}",
            oldTransaction.id,
            newWallet.name,
        )
    }

    private fun updateTransactionType(
        oldTransaction: WalletTransaction,
        newType: WalletTransactionType,
    ) {
        if (oldTransaction.type == newType) {
            logger.info("Transaction with id {} has the same type as before", oldTransaction.id)
            return
        }

        if (oldTransaction.status == WalletTransactionStatus.CONFIRMED) {
            when (oldTransaction.type) {
                WalletTransactionType.EXPENSE ->
                    incrementWalletBalance(
                        oldTransaction.wallet,
                        oldTransaction.amount.multiply(BigDecimal.TWO),
                    )
                WalletTransactionType.INCOME ->
                    decrementWalletBalance(
                        oldTransaction.wallet,
                        oldTransaction.amount.multiply(BigDecimal.TWO),
                    )
            }
        }

        oldTransaction.type = newType

        logger.info("Transaction with id {} type changed to {}", oldTransaction.id, newType)
    }

    private fun updateTransactionAmount(
        oldTransaction: WalletTransaction,
        newAmount: BigDecimal,
    ) {
        val diff = oldTransaction.amount.minus(newAmount)

        if (diff == BigDecimal.ZERO) {
            logger.info("Transaction with id {} has the same amount as before", oldTransaction.id)
            return
        }

        if (oldTransaction.status == WalletTransactionStatus.CONFIRMED) {
            val shouldIncrement =
                (oldTransaction.type == WalletTransactionType.EXPENSE && diff > BigDecimal.ZERO) ||
                    (oldTransaction.type == WalletTransactionType.INCOME && diff < BigDecimal.ZERO)

            if (shouldIncrement) {
                incrementWalletBalance(oldTransaction.wallet, diff.abs())
            } else {
                decrementWalletBalance(oldTransaction.wallet, diff.abs())
            }
        }

        oldTransaction.amount = newAmount

        logger.info("Transaction with id {} amount changed to {}", oldTransaction.id, newAmount)
    }

    private fun updateTransactionStatus(
        oldTransaction: WalletTransaction,
        newStatus: WalletTransactionStatus,
    ) {
        if (oldTransaction.status == newStatus) {
            logger.info("Transaction with id {} has the same status as before", oldTransaction.id)
            return
        }

        if (oldTransaction.status == WalletTransactionStatus.CONFIRMED && newStatus == WalletTransactionStatus.PENDING) {
            when (oldTransaction.type) {
                WalletTransactionType.EXPENSE -> incrementWalletBalance(oldTransaction.wallet, oldTransaction.amount)
                WalletTransactionType.INCOME -> decrementWalletBalance(oldTransaction.wallet, oldTransaction.amount)
            }
        } else {
            when (oldTransaction.type) {
                WalletTransactionType.EXPENSE -> decrementWalletBalance(oldTransaction.wallet, oldTransaction.amount)
                WalletTransactionType.INCOME -> incrementWalletBalance(oldTransaction.wallet, oldTransaction.amount)
            }
        }

        oldTransaction.status = newStatus

        logger.info("Transaction with id {} status changed to {}", oldTransaction.id, newStatus)
    }

    private fun revertTransferWalletBalances(
        oldTransfer: Transfer,
        newTransfer: Transfer,
    ) {
        val senderChanged = oldTransfer.senderWallet.id != newTransfer.senderWallet.id
        val receiverChanged = oldTransfer.receiverWallet.id != newTransfer.receiverWallet.id

        when {
            !senderChanged && !receiverChanged -> {
                val difference = newTransfer.amount.minus(oldTransfer.amount)
                when {
                    difference > BigDecimal.ZERO -> {
                        decrementWalletBalance(newTransfer.senderWallet, difference)
                        incrementWalletBalance(newTransfer.receiverWallet, difference)
                    }
                    difference < BigDecimal.ZERO -> {
                        incrementWalletBalance(newTransfer.senderWallet, difference.abs())
                        decrementWalletBalance(newTransfer.receiverWallet, difference.abs())
                    }
                }
            }
            senderChanged && !receiverChanged -> {
                incrementWalletBalance(oldTransfer.senderWallet, oldTransfer.amount)
                decrementWalletBalance(newTransfer.senderWallet, newTransfer.amount)

                val difference = newTransfer.amount.minus(oldTransfer.amount)
                when {
                    difference > BigDecimal.ZERO -> incrementWalletBalance(newTransfer.receiverWallet, difference)
                    difference < BigDecimal.ZERO -> decrementWalletBalance(newTransfer.receiverWallet, difference.abs())
                }
            }
            !senderChanged && receiverChanged -> {
                decrementWalletBalance(oldTransfer.receiverWallet, oldTransfer.amount)
                incrementWalletBalance(newTransfer.receiverWallet, newTransfer.amount)

                val difference = newTransfer.amount.minus(oldTransfer.amount)
                when {
                    difference > BigDecimal.ZERO -> decrementWalletBalance(newTransfer.senderWallet, difference)
                    difference < BigDecimal.ZERO -> incrementWalletBalance(newTransfer.senderWallet, difference.abs())
                }
            }
            else -> {
                incrementWalletBalance(oldTransfer.senderWallet, oldTransfer.amount)
                decrementWalletBalance(oldTransfer.receiverWallet, oldTransfer.amount)

                decrementWalletBalance(newTransfer.senderWallet, newTransfer.amount)
                incrementWalletBalance(newTransfer.receiverWallet, newTransfer.amount)
            }
        }
    }
}
